src/Storages/MergeTree/MergeTreeDataPartWriterWide.cpp

```diff
@@ -14,6 +15,75 @@ namespace
     constexpr auto DATA_FILE_EXTENSION = ".bin";
 }
 
+namespace
+{
+
+/// Get granules for block using index_granularity
+Granules getGranulesToWrite(const MergeTreeIndexGranularity & index_granularity, size_t block_rows, size_t current_mark, size_t rows_written_in_last_mark)
+{
+    if (current_mark >= index_granularity.getMarksCount())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Request to get granules from mark {} but index granularity size is {}", current_mark, index_granularity.getMarksCount());
+
+    Granules result;
+    size_t current_row = 0;
+    /// When our last mark is not finished yet and we have to write in rows into it
+    if (rows_written_in_last_mark > 0)
+    {
+        size_t rows_left_in_last_mark = index_granularity.getMarkRows(current_mark) - rows_written_in_last_mark;
+        size_t rest_rows = block_rows - current_row;
+        if (rest_rows < rows_left_in_last_mark)
+            result.emplace_back(Granule{
+                .start_row = current_row,
+                .granularity_rows = rows_left_in_last_mark,
+                .block_rows = rest_rows,
+                .mark_number = current_mark,
+                .mark_on_start = false, /// Don't mark this granule because we have already marked it
+            });
+        else
+            result.emplace_back(Granule{
+                .start_row = current_row,
+                .granularity_rows = rows_left_in_last_mark,
+                .block_rows = rows_left_in_last_mark,
+                .mark_number = current_mark,
+                .mark_on_start = false, /// Don't mark this granule because we have already marked it
```

Comment:
Maybe replace it with just `max(rows_left_in_last_mark, rest_rows)` for `block_rows`?

Reply:
yes, the code became more compact, but slightly more complex. But let's replace it.